System test runner MVP

ycombinator opened the pull request:
This PR fleshes out the implementation of the system tests runner. 

Initially it focusses on packages whose services can be spun up using Docker Compose but it has a provision to be extended later to include other types of services, e.g. cloud integration packages.

## Configuring the package/dataset for a system test
1. Every package must define how it's service must be spun up and torn down, by placing the appropriate files in the `_dev` folder located at the package's root folder.
   * Currently, only Docker Compose-based services are supported. 
   * For this, a `_dev/docker-compose.yml` file must be created. This file may define as many services as needed but the main one is determined by a naming convention: the name of the main service in the `docker-compose.yml` file must be the same name as the package. 

2. Additionally, every package dataset that wants to be system-tested must define a system test configuration file, `config.yml`, located in the `_dev/test/system` folder at the dataset's root folder.

For example, see [this PR](https://github.com/elastic/integrations/pull/263) setting up the above files for the `apache` package and its datasets.

## Running the system test
First, we must spin up the Elastic Stack and export environment variables.
```
$ elastic-package stack up --version 8.0.0-SNAPSHOT -d
$ eval $(elastic-package stack shellinit)
```

Then we can run the system test on a package (or specific dataset(s) within a package with the optional `--dataset` flag).
```
$ cd /path/to/some/package/root/folder
$ elastic-package test system
```

The above command will perform the following steps under the hood:

1. It will find the `_dev` folder at the package root and look inside it to decide how the service for this package must be spun up (and later torn down). 
   * Docker Compose-based services are started up as Docker containers on the same network as the Elastic Stack spun up with `elastic-package stack up`.

2. The service is spun up and certain **context variables** are registered. These variables become available to every dataset's test configuration file. Currently, the following context variables are available:
   * `{{ Service.STDOUT }}`: Available for Docker Compose-based services. Its value will be a file path available in the Agent container. This file will contain the contents of the service's STDOUT stream. It is useful for gathering logs from Docker Compose-based services as they write their logs to STDOUT by default.
   * `{{ Service.Ports.n.InternalPort }}`: Available for Docker-Compose based services. Its value will be the internal port available on the Docker network shared between the service container(s) and Elastic Stack containers. Note that the `n` here is a 0-based index, to allow for a service to listen on multiple internal ports.
   * `{{ Service.Port.InternalPort }}`: Available for Docker-Compose based services. It is simply a convenient alias for `{{ Service.Ports.0.InternalPort }}` as most Docker-Compose based services will only listen on a single internal port.
   * `{{ Service.Hostname }}`: Host name of service, as addressable from Agent container.

3. The dataset's **system test configuration file** is loaded into memory. Any context variables are substituted with their corresponding values. All other dataset and package variables omitted from the system test configuration file are populated from the package and dataset manifest files, respectively.

4. A **policy** is created for the dataset being tested in Ingest Manager. The policy is attached to the test Agent that is already spun up as part of the Elastic Stack and enrolled with Fleet. Within a minute or so the policy is sent to the Agent by Fleet.

5. Agent applies the policy and should start indexing data into the appropriate data stream(s) in Elasticsearch.

6. The system test runner checks if the expected data streams exist and have data (documents) in them.

ycombinator commented on internal/testrunner/runners/system/system.go:
```diff
@@ -3,29 +3,59 @@ package system
+	}
+
+	// Step 2. Setup test cluster (ES + Kibana + Agent).
+	// Step 2a. Register Agent with Fleet.
```
@mtojek I would like your advice here: what would be the best way to accomplish these two steps (2. and 2a.) here?

mtojek replied:
Step 2a:

It depends where the agent will be spawned. If it runs in a container then the enrollment should be performed automatically (it's in agent's Dockerfile).
I believe you have to interact with cluster to attach new container with observed product (like mysql or nginx) to the existing network, and detach it after tests.

Step 2:

I think you should assume that the cluster is already up and running, and the Agent in subscribed. You can pick up configuration options through environment variables (shellinit). I assume that the Agent is long-running and we just switch configuration inside.

ycombinator replied:
> I think you should assume that the cluster is already up and running, and the Agent in subscribed. You can pick up configuration options through environment variables (shellinit). I assume that the Agent is long-running and we just switch configuration inside.

This is the part I am most interested in. Is there some way I can _ensure_ that the cluster is up, rather than assuming it is? Are there any functions/methods I could call from here to make that happen?

> I believe you have to interact with cluster to attach new container with observed product (like mysql or nginx) to the existing network, and detach it after tests.

Cool, this is along the lines of what I was hoping for. For this I would need to know the name of the network. Is there some function/method I could call from here to get that (related to the previous paragraph, I suppose)?

mtojek replied:
> This is the part I am most interested in. Is there some way I can ensure that the cluster is up, rather than assuming it is? Are there any functions/methods I could call from here to make that happen idempotently?

I think it' s `cluster.BootUp()`, the only issue we need to figure out is waiting for Kibana. I believe you start with the assumption that the cluster is up and fail in case it's not.

>  For this I would need to know the name of the network. Is there some function/method I could call from here to get that (related to the previous paragraph, I suppose)?

There is no such function at the moment, as it wasn't required, but I suppose we can enforce any name store it in `.env` file in `.elastic-package`.

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-64/detail/PR-64/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-64/detail/PR-64/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-64/detail/PR-64/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-64/detail/PR-64/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-package_64.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #64 opened]`

* Start Time: 2020-08-18T22:40:22.813+0000

* Duration: 2 min 9 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: `Check`
    * Description: make check
    
    * Duration: 0 min 16 sec
    
    * Start Time: 2020-08-18T22:42:14.055+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/elastic-package/pipelines/PR-64/runs/1/steps/114/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-08-18T22:41:51.169Z]  > git --version # timeout=10
[2020-08-18T22:41:51.175Z] fatal: bad object 06c62f097a3109e82d977b3292df1e77ce92ff26
[2020-08-18T22:41:51.039Z]  > git config core.sparsecheckout # timeout=10
[2020-08-18T22:41:51.043Z]  > git checkout -f 155ea62e882ea2443f22a082184de9a256295a02 # timeout=15
[2020-08-18T22:41:51.064Z]  > git remote # timeout=10
[2020-08-18T22:41:51.071Z]  > git config --get remote.origin.url # timeout=10
[2020-08-18T22:41:51.075Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-08-18T22:41:51.079Z]  > git merge e17163244e0aab406e5efa3f00be3c22b50e6e51 # timeout=10
[2020-08-18T22:41:51.088Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-08-18T22:41:51.099Z]  > git config core.sparsecheckout # timeout=10
[2020-08-18T22:41:51.103Z]  > git checkout -f 155ea62e882ea2443f22a082184de9a256295a02 # timeout=15
[2020-08-18T22:41:51.122Z]  > git rev-parse --verify HEAD # timeout=10
[2020-08-18T22:41:51.131Z] Resetting working tree
[2020-08-18T22:41:51.131Z]  > git reset --hard # timeout=10
[2020-08-18T22:41:51.137Z]  > git clean -fdx # timeout=10
[2020-08-18T22:41:52.170Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-08-18T22:41:52.921Z] + git fetch https://****:****@github.com/elastic/elastic-package.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-08-18T22:41:53.253Z] Archiving artifacts
[2020-08-18T22:41:54.023Z] + git rev-parse HEAD
[2020-08-18T22:41:54.419Z] + git rev-parse HEAD
[2020-08-18T22:41:54.758Z] + git rev-parse origin/pr/64
[2020-08-18T22:41:54.817Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-08-18T22:41:55.443Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-08-18T22:41:56.296Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/elastic-package/pulls/64/reviews return 0 elements
[2020-08-18T22:41:56.373Z] [INFO] githubPrCheckApproved: Title: Fleshing out system test runner a bit - User: ycombinator - Author Association: MEMBER
[2020-08-18T22:41:56.989Z] Stashed 161 file(s)
[2020-08-18T22:41:57.073Z] Running in /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package
[2020-08-18T22:41:58.039Z] + gvm 1.14.2
[2020-08-18T22:42:03.351Z] export GOROOT="/var/lib/jenkins/workspace/Beats_elastic-package_PR-64/.gvm/versions/go1.14.2.linux.amd64"
[2020-08-18T22:42:03.352Z] export PATH="/var/lib/jenkins/workspace/Beats_elastic-package_PR-64/.gvm/versions/go1.14.2.linux.amd64/bin:$PATH"
[2020-08-18T22:42:03.718Z] + go get -u github.com/magefile/mage
[2020-08-18T22:42:06.024Z] + go get -u github.com/elastic/go-licenser
[2020-08-18T22:42:07.339Z] + go get -u golang.org/x/tools/cmd/goimports
[2020-08-18T22:42:13.024Z] + go get -u github.com/jstemmer/go-junit-report
[2020-08-18T22:42:14.032Z] Running in /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package
[2020-08-18T22:42:14.356Z] + make check
[2020-08-18T22:42:14.356Z] go get -ldflags "-X github.com/elastic/elastic-package/internal/version.CommitHash=`git describe --always --long --dirty` -X github.com/elastic/elastic-package/internal/version.BuildTime=`date +%s`" \
[2020-08-18T22:42:14.356Z]     github.com/elastic/elastic-package
[2020-08-18T22:42:14.619Z] go: downloading github.com/elastic/package-spec/code/go v0.0.0-20200817225106-d939a82ed059
[2020-08-18T22:42:14.619Z] go: downloading golang.org/x/oauth2 v0.0.0-20200107190931-bf48bf16ab8d
[2020-08-18T22:42:14.619Z] go: downloading github.com/google/go-github/v32 v32.1.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/pkg/errors v0.9.1
[2020-08-18T22:42:14.619Z] go: downloading gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c
[2020-08-18T22:42:14.619Z] go: downloading github.com/spf13/cobra v1.0.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/go-git/go-git/v5 v5.1.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/AlecAivazis/survey/v2 v2.1.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/Masterminds/semver v1.5.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/go-git/go-billy/v5 v5.0.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/rakyll/statik v0.1.7
[2020-08-18T22:42:14.619Z] go: downloading golang.org/x/crypto v0.0.0-20200622213623-75b288015ac9
[2020-08-18T22:42:14.619Z] go: downloading github.com/Masterminds/semver/v3 v3.1.0
[2020-08-18T22:42:14.619Z] go: downloading github.com/inconshreveable/mousetrap v1.0.0
[2020-08-18T22:42:14.884Z] go: downloading github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
[2020-08-18T22:42:14.884Z] go: downloading github.com/magefile/mage v1.10.0
[2020-08-18T22:42:14.884Z] go: downloading golang.org/x/text v0.3.2
[2020-08-18T22:42:14.884Z] go: downloading github.com/spf13/pflag v1.0.3
[2020-08-18T22:42:14.884Z] go: downloading golang.org/x/net v0.0.0-20200625001655-4c5254603344
[2020-08-18T22:42:14.884Z] go: downloading github.com/mitchellh/go-homedir v1.1.0
[2020-08-18T22:42:14.884Z] go: downloading github.com/google/go-querystring v1.0.0
[2020-08-18T22:42:14.884Z] go: downloading github.com/go-git/gcfg v1.5.0
[2020-08-18T22:42:14.884Z] go: downloading google.golang.org/appengine v1.4.0
[2020-08-18T22:42:14.884Z] go: downloading github.com/imdario/mergo v0.3.9
[2020-08-18T22:42:15.149Z] go: downloading github.com/sergi/go-diff v1.1.0
[2020-08-18T22:42:15.149Z] go: downloading github.com/mattn/go-isatty v0.0.8
[2020-08-18T22:42:15.149Z] go: downloading github.com/kevinburke/ssh_config v0.0.0-20190725054713-01f96b0aa0cd
[2020-08-18T22:42:15.149Z] go: downloading github.com/xanzy/ssh-agent v0.2.1
[2020-08-18T22:42:15.149Z] go: downloading github.com/mgutz/ansi v0.0.0-20170206155736-9520e82c474b
[2020-08-18T22:42:15.149Z] go: downloading golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd
[2020-08-18T22:42:15.149Z] go: downloading github.com/jbenet/go-context v0.0.0-20150711004518-d14ea06fba99
[2020-08-18T22:42:15.149Z] go: downloading gopkg.in/warnings.v0 v0.1.2
[2020-08-18T22:42:15.149Z] go: downloading github.com/mattn/go-colorable v0.1.2
[2020-08-18T22:42:15.149Z] go: downloading github.com/emirpasic/gods v1.12.0
[2020-08-18T22:42:15.149Z] go: downloading github.com/golang/protobuf v1.3.2
[2020-08-18T22:42:27.469Z] gofmt -s -w .
[2020-08-18T22:42:27.469Z] go get -u golang.org/x/lint/golint
[2020-08-18T22:42:27.469Z] go: downloading golang.org/x/lint v0.0.0-20200302205851-738671d3881b
[2020-08-18T22:42:27.469Z] go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20200302205851-738671d3881b
[2020-08-18T22:42:27.469Z] go: downloading golang.org/x/lint v0.0.0-20190313153728-d0100b6bd8b3
[2020-08-18T22:42:27.469Z] go: downloading golang.org/x/tools v0.0.0-20190311212946-11955173bddd
[2020-08-18T22:42:27.469Z] go: golang.org/x/lint upgrade => v0.0.0-20200302205851-738671d3881b
[2020-08-18T22:42:27.469Z] go: golang.org/x/tools upgrade => v0.0.0-20200818005847-188abfa75333
[2020-08-18T22:42:27.469Z] go: downloading golang.org/x/tools v0.0.0-20200818005847-188abfa75333
[2020-08-18T22:42:28.046Z] go list ./... | xargs -n 1 golint -set_exit_status
[2020-08-18T22:42:29.976Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package/internal/testrunner/runners/system/service/compose.go:3:6: exported type DockerComposeRunner should have comment or be unexported
[2020-08-18T22:42:29.976Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package/internal/testrunner/runners/system/service/compose.go:7:1: exported function NewDockerComposeRunner should have comment or be unexported
[2020-08-18T22:42:29.976Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package/internal/testrunner/runners/system/service/compose.go:13:1: exported method DockerComposeRunner.SetUp should have comment or be unexported
[2020-08-18T22:42:29.976Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package/internal/testrunner/runners/system/service/compose.go:13:40: exported method SetUp returns unexported type service.ctx, which can be annoying to use
[2020-08-18T22:42:29.976Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package/internal/testrunner/runners/system/service/compose.go:17:1: exported method DockerComposeRunner.TearDown should have comment or be unexported
[2020-08-18T22:42:29.976Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-64/src/github.com/elastic/elastic-package/internal/testrunner/runners/system/service/factory.go:10:2: exported var ErrNotFound should have comment or be unexported
[2020-08-18T22:42:29.976Z] Found 6 lint suggestions; failing.
[2020-08-18T22:42:29.976Z] Makefile:9: recipe for target 'lint' failed
[2020-08-18T22:42:29.976Z] make: *** [lint] Error 123
[2020-08-18T22:42:30.650Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_elastic-package_PR-64
[2020-08-18T22:42:30.783Z] [INFO] getVaultSecret: Getting secrets
[2020-08-18T22:42:30.864Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-08-18T22:42:31.670Z] + chmod 755 generate-build-data.sh
[2020-08-18T22:42:31.670Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-64/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-64/runs/1 FAILURE 128583
[2020-08-18T22:42:31.670Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-64/runs/1/steps/?limit=10000 -o steps-info.json
[2020-08-18T22:42:31.920Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-64/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-08-18T22:42:31.920Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-64/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

ycombinator commented:
@mtojek I just built `elastic-package` from this PR and tried running `elastic-package test system` from the `apache` package folder. I got this error:

```
$ elastic-package test system
system run /Users/shaunak/development/github/integrations/packages/apache/dataset/access/_dev/test/system
Custom build packages directory found: /Users/shaunak/development/github/integrations/build/integrations
Building package-registry
Step 1/3 : FROM docker.elastic.co/package-registry/distribution:snapshot
ERROR: Service 'package-registry' failed to build: Get https://docker.elastic.co/v2/package-registry/distribution/manifests/snapshot: unauthorized: authentication required
Error: error running package system tests: could not setup test cluster: building docker images failed: running command failed: exit status 1
```

Any idea what causes this / how a user of `elastic-package` is expected to authenticate?

ycombinator commented:
I was able to resolve the issue in my previous comment. It was a one-off issue, so nothing we need to worry about in general with `elastic-package` users.

mtojek commented on internal/testrunner/testrunner.go:
```diff
@@ -55,12 +76,12 @@ func RegisterRunner(testType TestType, runFunc RunFunc) {
 
 // Run method delegates execution to the registered test runner, based on the test type.
-func Run(testType TestType, testFolderPath string) error {
+func Run(testType TestType, testFolder TestFolder, packageRootPath string) error {
```
I think we can sync up both branches.

I introduced a similar concept of `TestOptions`: https://github.com/elastic/elastic-package/blob/c3fd692e9923f4160b11b3eac569b76014552934/internal/testrunner/testrunner.go to easily add more arguments here.

mtojek commented on internal/testrunner/runners/system/system.go:
```diff
@@ -2,30 +2,134 @@ package system
+		return errors.Wrap(err, "could not create service runner")
+	}
+
+	ctxt := common.MapStr{}
```
I think it's still early and there is no implementation using `MapStr`. Do you plan to use the MapStr to flatten JSON object representation? If so, I wonder if standars `encoding/json` is sufficient.

mtojek replied:
Also, this is the reason why I don't flatten the expected results in the pipeline tests. Maybe I should introduce it there.

ycombinator replied:
I'm not 100% sure yet how this usage is going to evolve. My current thinking is to use `ctxt` to pass some arbitrary key-value pair data between the various steps of testing but this is something I'm experimenting with.

For test results, since those are consistently going to be present for each test run, I probably wouldn't use `ctxt` but something else. I haven't gotten that far yet so I'm not sure yet.

mtojek commented on internal/kibana/ingestmanager/client.go:
```diff
@@ -0,0 +1,16 @@
+package ingestmanager
+
+type Client struct {
```
I used a dedicated client for Elasticsearch. Maybe there is something similar for Kibana?

ycombinator replied:
I have not found a reliable one so far. I suspect, over time, as Kibana APIs become more "public" we will see one emerge. At that point we can use it here.

ycombinator replied:
BTW, looks like there is some community interest in this too: https://github.com/elastic/go-elasticsearch/issues/182

mtojek replied:
Oh, nice :)

mtojek commented on internal/testrunner/runners/system/system.go:
```diff
@@ -2,30 +2,134 @@ package system
 
-func init() {
-	testrunner.RegisterRunner(TestType, Run)
+func getStackSettingsFromEnv() stackSettings {
```
I exposed environment variables here: https://github.com/elastic/elastic-package/blob/c3fd692e9923f4160b11b3eac569b76014552934/internal/elasticsearch/client.go Does it make sense to use them here?

ycombinator replied:
I'm open to it but I'm not sure how to do that. Can you elaborate what should change here?

mtojek replied:
I suppose it would be merging the other PR and adjust the implementation here to use environment variables defined in a single place (e.g. close to the stack).

ycombinator replied:
Okay, I plan to bring the changes from the other PR into this one today. Let's see how things look at that point.

mtojek commented on internal/install/static_snapshot_yml.go:
```diff
@@ -69,4 +69,31 @@ services:
+    volumes:
+    - type: bind
+      source: ../tmp/
+      target: /tmp/service_logs/
```
Ok, so with this solution, we lose the opportunity to test on real filesystem and have log files distributed across the entire partition. We might need a comment somewhere about this limitation or a task for next iteration (maybe we can solve it somehow).

ycombinator replied:
Not necessarily. There is no requirement for log files paths in the dataset configuration to be restricted to `/tmp/service_logs`. So, for example, when we write a test for the `system` package, we can refer to standard locations for system log files assuming the Agent is running some flavor of Linux.

mtojek commented on internal/testrunner/runners/system/system.go:
```diff
@@ -5,31 +5,364 @@
+		return errors.Wrap(err, "could not setup service")
+	}
+
+	// Step 2. Configure package (single data stream) via Ingest Manager APIs.
```
Actually you don't use any "install" API like the one that is used in integration tests of package-registry. It looks like it's not required to be involved in the process, so I wonder what does it do actually? For reference: https://github.com/elastic/package-registry/blob/master/main_integration_test.go#L82

Maybe it would be fine to call it for safety reasons, to verify if the package can be installed?

ycombinator replied:
That's because it is being called by Agent. See https://github.com/elastic/elastic-package/pull/64/files#diff-906969eaf1b600ec57ae5e30f11d1a3dR87, https://github.com/elastic/beats/blob/2a0e0998243db56d0bfd681ff3cb06183166bdd5/dev-tools/packaging/templates/docker/docker-entrypoint.elastic-agent.tmpl#L64, and https://github.com/elastic/beats/blob/master/dev-tools/packaging/templates/docker/docker-entrypoint.elastic-agent.tmpl#L16.

mtojek replied:
Sorry, I misguided you. I was thinking about the reference to this API method: https://github.com/elastic/package-storage/blob/production/testing/main_integration_test.go#L108

ycombinator replied:
Hmmm, good question — I wonder how I'm able to get things working even without calling this API. 🤔 Let's leave it out of this PR for now and add it if we need it in a follow up PR.

mtojek commented on internal/testrunner/runners/system/system.go:
```diff
@@ -5,31 +5,364 @@
+
+	logger.Info("creating test policy...")
+	p := ingestmanager.Policy{
+		Name:        fmt.Sprintf("ep-test-system-%s-%s", r.testFolder.Package, r.testFolder.Dataset),
```
I didn't see you applied any timestamps here. I wonder if it's not better to include some. Consider a scenario in which developer stops the test runner in the middle and for some reason the cleanup never happened or failed. We can prevent some transient errors in the future.

ycombinator replied:
Yes, we could do that. Another option would be to not use timestamps so we have the same identifiers for each test run. Then we clean up resources _before_ (and after) the test run starts, in case any resources were left from transient failures from a previous run.

Either way, I agree we need to account for orphan resources due to transient failures from previous runs.

WDYT?

mtojek replied:
> Then we clean up resources before (and after) the test run starts, in case any resources were left from transient failures from a previous run.

As long as we don't have intention to run tests in parallel, this can work pretty well.

ycombinator replied:
Actually, what I suggested above is not trivial to implement for policies as their IDs are generated by Kibana. We could retrieve all policies and delete ones that match a certain naming convention, but let's worry about that in a follow up PR.

For now, I took your suggestion and added a timestamp to the policy name (via 9fde2df).

mtojek commented on internal/testrunner/runners/system/servicedeployer/compose.go:
```diff
@@ -0,0 +1,177 @@
+	serviceContainer := fmt.Sprintf("%s_%s_1", r.project, serviceName)
+	ctxt.Put("Service.Hostname", serviceContainer)
+
+	// Redirect service container's STDOUT and STDERR streams to files in local logs folder
```
I wonder if we can simplify this - we still have a volume with log files mounted, so maybe don't depend specifically on stdout and stderr, but just share a volume between agent and the service. I'm fine with discussing this in a different issue.

ycombinator replied:
I think I'm missing something here. Let's use the `apache` container as an example. Let's say we have a shared volume between this container and the `elastic-agent` container. I assume we want Apache log files to be placed by the `apache` container into this volume. But how/where exactly do we make this happen?

ycombinator replied:
I just read your overall PR review comment in which you mention this:

> In the service configuration, we define paths to be exposed as mounts...

I think that answers my question above, "how/where exactly do we make this happen?" We would define them in the `docker-compose.yml` file for the service. Let me try this out and see if we can make this work.

mtojek commented:
Ok, I went through the source code, left couple of high level comments. I really like this, it's not full-blown testing framework, but a simple runner. Nicely done! In general I'm for merging a more or less experimental version of the feature and simply not enabling it in the Integrations repository.

My only concerns are around "recording" stdout and stderr of the container instead of sharing a mounted volume between the agent and deployed service. A possible solution to be considered:

1. In the service configuration, we define paths to be exposed as mounts (if there are multiple directories with logs).
2. Service and agent are deployed on demand only for the lifetime of the test (always in pair).
3. They share multiple volumes with logs.

Re 2: I understand it might be hard to determine when to start or stop the service&agent while interacting with multiple test cases, as enrolling and unenrolling every time is time-consuming. On the other hand, we have guarantee that the agent picked up the right configuration.

ycombinator commented on internal/install/static_snapshot_yml.go:
```diff
@@ -69,4 +69,31 @@ services:
+    - "KIBANA_HOST=http://kibana:5601"
+    volumes:
+    - type: bind
+      source: ../tmp/
```
Sure, I'll call it `service_logs` so it matches the name of the folder in the agent container (`/tmp/service_logs`).

ycombinator replied:
Fixed in fcec6d1.

ycombinator commented on internal/install/static_snapshot_yml.go:
```diff
@@ -69,4 +69,31 @@ services:
+    - "FLEET_ENROLL_INSECURE=1"
+    - "FLEET_SETUP=1"
+    - "KIBANA_HOST=http://kibana:5601"
+    volumes:
```
I did not, but will do.

ycombinator replied:
Hmmm, interesting. It seems to work for `package-registry` and `elasticsearch` but not for `kibana`:

### Starting up only `package-registry` ✔️ 

```
$ elastic-package stack up --version 8.0.0-SNAPSHOT -d -v -s package-registry
2020/09/01 10:34:28 DEBUG Enable verbose logging
Boot up the Elastic stack
2020/09/01 10:34:28 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack build package-registry package-registry_is_ready
2020/09/01 10:34:53 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack down
2020/09/01 10:34:54 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack up -d package-registry package-registry_is_ready
Done
```
```
$ docker ps
CONTAINER ID        IMAGE                                    COMMAND                  CREATED             STATUS                    PORTS                      NAMES
2ff3fbee2ecb        elastic-package-stack_package-registry   "./package-registry …"   18 seconds ago      Up 17 seconds (healthy)   127.0.0.1:8080->8080/tcp   elastic-package-stack_package-registry_1
```

### Starting up only `elasticsearch` ✔️ 

```
$ elastic-package stack up --version 8.0.0-SNAPSHOT -d -v -s elasticsearch
2020/09/01 10:41:34 DEBUG Enable verbose logging
Boot up the Elastic stack
2020/09/01 10:41:34 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack build elasticsearch elasticsearch_is_ready
2020/09/01 10:41:34 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack down
2020/09/01 10:41:35 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack up -d elasticsearch elasticsearch_is_ready
Done
```
```
$ docker ps
CONTAINER ID        IMAGE                                                          COMMAND                  CREATED              STATUS                        PORTS                                NAMES
826522aaf207        docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT   "/tini -- /usr/local…"   About a minute ago   Up About a minute (healthy)   127.0.0.1:9200->9200/tcp, 9300/tcp   elastic-package-stack_elasticsearch_1
```

### Starting up only `kibana` ❌ 

```
$ elastic-package stack up --version 8.0.0-SNAPSHOT -d -v -s kibana
2020/09/01 10:36:42 DEBUG Enable verbose logging
Boot up the Elastic stack
2020/09/01 10:36:42 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack build
2020/09/01 10:37:06 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack down
2020/09/01 10:37:07 DEBUG running command: /usr/local/bin/docker-compose -f /Users/shaunak/.elastic-package/stack/snapshot.yml -p elastic-package-stack up -d
Done
```
```
$ docker ps
CONTAINER ID        IMAGE                                                          COMMAND                  CREATED              STATUS                        PORTS                                NAMES
f00fbb57d861        docker.elastic.co/beats/elastic-agent:8.0.0-SNAPSHOT           "/usr/local/bin/dock…"   40 seconds ago       Up 39 seconds (healthy)                                            elastic-package-stack_elastic-agent_1
a3b35e4a30ad        docker.elastic.co/kibana/kibana:8.0.0-SNAPSHOT                 "/usr/local/bin/dumb…"   About a minute ago   Up About a minute (healthy)   127.0.0.1:5601->5601/tcp             elastic-package-stack_kibana_1
6f35f3833863        docker.elastic.co/elasticsearch/elasticsearch:8.0.0-SNAPSHOT   "/tini -- /usr/local…"   2 minutes ago        Up 2 minutes (healthy)        127.0.0.1:9200->9200/tcp, 9300/tcp   elastic-package-stack_elasticsearch_1
0c4cd0260cf1        elastic-package-stack_package-registry                         "./package-registry …"   2 minutes ago        Up 2 minutes (healthy)        127.0.0.1:8080->8080/tcp             elastic-package-stack_package-registry_1
```

Wonder why that is? 🤔 Investigating...

ycombinator replied:
Oh, it's because of this check: https://github.com/elastic/elastic-package/blob/ba7825e1b010cab16705f47c5c6d1fad332edd52/internal/stack/boot.go#L186-L188

I will update that to check for `elastic-agent` now as that is the "root" service in the dependency tree now.

ycombinator replied:
Updated in 730738123a978601e4e871e2958d16cfd5f8d18a.

ycombinator commented on internal/install/static_snapshot_yml.go:
```diff
@@ -69,4 +69,31 @@ services:
+      kibana:
+        condition: service_healthy
+    healthcheck:
+      test: ["CMD", "grep", "Agent is starting", "/usr/share/elastic-agent/elastic-agent.log"]
```
I'm not sure I understand this point. Are you suggesting we tell Agent to write its log files to `/tmp` instead of where it writes them by default and change the health check here somehow?

mtojek replied:
Sorry for the confusion, it's a rather a nice-to-have feature to also observe programmatically logs of elastic-agent, filebeat and metricbeat while testing. Many configuration errors are reported deep down by filebeat or metricbeat's internals.
Logs can be collected in the same single directory and exported as volume.

ycombinator replied:
Thanks for explaining. Let's handle this in a follow up PR.

ycombinator commented on internal/testrunner/runners/system/config.go:
```diff
@@ -0,0 +1,57 @@
+	return &c, nil
+}
+
+func applyContext(ctxt common.MapStr, data []byte) ([]byte, error) {
```
Okay, I will add some comments explaining why.

ycombinator replied:
I took a crack at this in 38c62ae. Let me know if you want further changes.

ycombinator commented on internal/testrunner/runners/system/servicedeployer/compose.go:
```diff
@@ -0,0 +1,177 @@
+		return errors.Wrapf(err, "could not delete STDERR file: %s", r.stderrFilePath)
+	}
+
+	if err := r.stdout.Close(); err != nil {
```
Yeah, good idea to move this to a `defer`. Will do!

ycombinator replied:
Fixed in f24d666.

ycombinator commented on internal/testrunner/runners/system/servicedeployer/factory.go:
```diff
@@ -0,0 +1,27 @@
+
+// Factory chooses the appropriate service runner for the given package, depending
+// on service configuration files defined in the package.
+func Factory(packageRootPath string) (Runner, error) {
```
I like it! Will change.

ycombinator replied:
Done in fcae0b7. Let me know what you think.

ycombinator commented on internal/testrunner/runners/system/system.go:
```diff
@@ -5,31 +5,364 @@
+	}
+
+	ctxt := common.MapStr{}
+	ctxt.Put("service.name", r.testFolder.Package)
```
Yeah, I think we can do a hybrid here. I will change it to a struct and add any "standard" properties that we expect to exist all the time. But I will also add a `common.MapStr` field to the struct for arbitrary key/values.

ycombinator replied:
Actually I went all the way and made the context a struct. See c0c672a.

ycombinator commented on internal/testrunner/runners/system/system.go:
```diff
@@ -5,31 +5,364 @@
+	}
+
+	// Delete old data
+	esClient, err := elasticsearch.Client()
```
Good idea to share one client across multiple invocations of the test runner. Will change.

ycombinator replied:
Done in 7ee7bd9064a73f3098a8b9e1f54cfdf6b0e351dc. I used it in the Pipeline Test Runner as well so you might want to take a look at that.

ycombinator commented on internal/kibana/ingestmanager/agents.go:
```diff
@@ -0,0 +1,52 @@
+package ingestmanager
```
Renamed `agents.go` and `policies.go` in cb87e8f.

ycombinator commented:
@mtojek I'm stopping for the day so posting an update here about addressing your feedback so far.

I've addressed _most_ of your feedback by now. The main item remaining is to explore sharing logs between the service container and agent via a volume instead of redirecting STDOUT and STDERR: https://github.com/elastic/elastic-package/pull/64#discussion_r481066466. I will look into this tomorrow.

Also, while I look into the shared volume implementation, I have questions about a couple of your suggestions: https://github.com/elastic/elastic-package/pull/64#discussion_r481281429 and https://github.com/elastic/elastic-package/pull/64#discussion_r481295862.

mtojek commented on internal/common/context.go:
```diff
@@ -0,0 +1,45 @@
+// ServiceContext encapsulates context that is both available to a ServiceDeployer and
+// populated by a DeployedService. The fields in ServiceContext may be used in handlebars
+// templates in system test configuration files, for example: {{ Hostname }}.
+type ServiceContext struct {
```
I wonder if the `ServiceContext` implementation should be stored close to the `ServiceDeployer` implementation, in the `servicedeployer` package. If the package name doesn't fit well, what do you think about `deployment` or something similar.

ycombinator replied:
I agree! I think I originally had it in the `servicedeployer` package but then ran into some import cycle issues, which is when I moved it out into `common`. But I just moved it back into `servicedeployer` now (via 65bf419) and I was able to build `elastic-package`.

mtojek reviewed (commented):
Yeah, I like the way it looks now, my only concern is related to the place where the context is stored.

Also, it seems that CI failed with:
```
[2020-09-01T23:15:55.736Z] + elastic-package stack up -d
[2020-09-01T23:15:55.736Z] elastic-package has been installed.
[2020-09-01T23:15:55.736Z] Boot up the Elastic stack
[2020-09-01T23:19:17.373Z] Error: booting up the stack failed: running docker-compose failed: running command failed: running Docker Compose up command failed: exit status 1
```

For debugging purposes, you can enable verbose mode in this test (currently it doesn't say too much about the root cause).

Once this PR is merged, we can focus on creating followup issues and prioritize them.

ycombinator commented on internal/testrunner/runners/system/servicedeployer/deployed_service.go:
```diff
@@ -0,0 +1,21 @@
+	TearDown() error
+
+	// GetContext returns the current context from the service.
+	GetContext() common.ServiceContext
```
I like the parallel with `SetContext` so I would prefer to leave it as `GetContext`.

mtojek replied:
Sure, let's leave it :)

For reference - I was referring to https://golang.org/doc/effective_go.html#Getters

ycombinator replied:
Aha! Thanks for the reference. I will rename it :)

ycombinator replied:
Fixed in 8c1c071.

ycombinator commented:
CI failure should be fixed by 326120ec4a736ab3c072fffef22037a14ffecd25.

mtojek commented:
> CI failure should be fixed by [326120e](https://github.com/elastic/elastic-package/commit/326120ec4a736ab3c072fffef22037a14ffecd25).

Out of curiosity, why did it fail?

ycombinator commented:
> Out of curiosity, why did it fail?

Before 326120e, we were using `7.9.0` as the default version of the stack. In this PR we introduce the Elastic Agent in the Docker Compose file. Elastic Agent 7.9.0's Docker image did not have the ability to set the `--insecure` flag when talking to Kibana. This ability was added in 7.9.1 (https://github.com/elastic/beats/pull/20713), so I needed to bump up the default version to 7.9.1.

mtojek commented:
Clear now! Thanks for explanation.

ycombinator commented:
There is one more issue we need to deal with w.r.t. the default stack version and this PR. As of now in this PR the default stack version is set to 7.9.1. However, this PR calls Kibana APIs that are only available starting 7.10.0 — actually these APIs are available in 7.9.0 as well but they use different identifiers in that version, e.g. "config" instead of "policy".

So what should we do about this? We could bump the default stack version to 7.10.0-SNAPSHOT _or_ we could rename the API identifiers to the old ones. Personally I lean towards the former option as we know users are unlikely to use `elastic-package` with 7.9.0 of the stack. What is your preference, @mtojek?

mtojek commented on internal/compose/compose.go:
```diff
@@ -67,7 +136,7 @@ func (p *Project) Down(opts CommandOptions) error {
 	args = append(args, opts.ExtraArgs...)
 
-	if err := runDockerComposeCmd(args, opts.Env); err != nil {
+	if err := p.runDockerComposeCmd(args, opts.Env, nil); err != nil {
```
I'm admit I'm allergic to passing nil values as it tends to evolve into WinAPI "NULL" syndrome: `call(a, b, 1, 1, NULL, NULL, NULL, NULL, NULL, true, NULL, 1, NULL)` :)

I think you can repack all these args into `dockerComposeOptions`? Less refactoring in the future.

ycombinator replied:
Done in e0f21d7.

mtojek reviewed (approved):
> So what should we do about this? We could bump the default stack version to 7.10.0-SNAPSHOT or we could rename the API identifiers to the old ones. Personally I lean towards the former option as we know users are unlikely to use elastic-package with 7.9.0 of the stack. What is your preference, @mtojek?

I'm for the option that will be less painful for developers (elastic-package users), so let's bump it up to the 7.10.0-SNAPSHOT.

Anyway, I think is in the good shape to be merged. I left a comment about passing options, but you can postpone or ignore it :) LGTM!

mtojek commented:
@ycombinator let's get it merge and see in the wild how does it operate on real packages.

ycombinator commented:
> Anyway, I think is in the good shape to be merged.

Thanks. There are quite a few suggestions/ideas in this PR that need follow up PRs:
- Volume mounting: https://github.com/elastic/elastic-package/pull/64#discussion_r481066466
- Observing logs of Agent and underlying Beats while testing: https://github.com/elastic/elastic-package/pull/64#discussion_r480965636
- Calling the package installation API?: https://github.com/elastic/elastic-package/pull/64#discussion_r485172111
- Deleting old test assets before running a test: https://github.com/elastic/elastic-package/pull/64#discussion_r485200235

I will file issues for each of these after I merge this PR and update this comment with references.

mtojek commented:
Works for me! We can easily share them.

ycombinator commented:
Investigating the latest CI failure...

ycombinator commented:
Updated the health check for the Elastic Agent service in d1886e7 to check for the expected log entry across multiple log files, in case the logs had rotated before the health check started running. This should fix CI.
